Refresh.GameServer/Database/GameDatabaseContext.Activity.cs

```diff
@@ -1,23 +1,103 @@
 using JetBrains.Annotations;
+using MongoDB.Bson;
+using Refresh.GameServer.Services;
 using Refresh.GameServer.Types.Activity;
 using Refresh.GameServer.Types.Levels;
+using Refresh.GameServer.Types.Relations;
+using Refresh.GameServer.Types.UserData;
 
 namespace Refresh.GameServer.Database;
 
 public partial class GameDatabaseContext // Activity
 {
     [Pure]
-    public DatabaseList<Event> GetRecentActivity(int count, int skip, long timestamp, long endTimestamp) => 
-        new(this._realm.All<Event>()
-            .Where(e => e.Timestamp < timestamp && e.Timestamp >= endTimestamp)
-            .OrderByDescending(e => e.Timestamp), skip, count);
+    public DatabaseList<Event> GetUserRecentActivity(
+        ActivityQueryParameters parameters,
+        FriendStorageService? friendService = null
+    )
+    {
+        IEnumerable<Event> query = this.GetRecentActivity(parameters, friendService);
+
+        if (parameters.User != null)
+        {
+            List<ObjectId?> favouriteUsers = parameters.User.UsersFavourited.AsEnumerable().Select(f => (ObjectId?)f.UserToFavourite.UserId).ToList();
+            List<ObjectId?>? userFriends = friendService?.GetUsersFriends(parameters.User, this)?.Select(u => (ObjectId?)u.UserId).ToList();
+
+            query = query.Where(e =>
+                e.User.UserId == parameters.User.UserId ||
+                e.StoredObjectId == parameters.User.UserId ||
+                favouriteUsers.Contains(e.User.UserId) ||
+                favouriteUsers.Contains(e.StoredObjectId) ||
+                (userFriends?.Contains(e.User.UserId) ?? false) ||
+                (userFriends?.Contains(e.StoredObjectId) ?? false) ||
+                this.GetLevelById(e.StoredSequentialId ?? int.MaxValue)?.Publisher?.UserId == parameters.User.UserId ||
+                e.EventType == EventType.Level_TeamPick ||
+                e.EventType == EventType.User_FirstLogin
+            );
+        }
+        
+        return new(query.OrderByDescending(e => e.Timestamp), parameters.Skip, parameters.Count);
+    }
+
+    private IEnumerable<Event> GetRecentActivity(ActivityQueryParameters parameters,
+        FriendStorageService? friendService = null)
+    {
+        if (parameters.Timestamp == 0) 
+            parameters.Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+        
+        IEnumerable<Event> query = this._realm.All<Event>()
+            .Where(e => e.Timestamp < parameters.Timestamp && e.Timestamp >= parameters.EndTimestamp)
+            .AsEnumerable();
+
+        if (parameters is { ExcludeMyLevels: true, User: not null })
```

Comment:
This style of syntax is a little untrustworthy. Is this function backed by unit tests?

Reply:
Its not backed by unit tests to my knowledge, but this logic was generated using Rider's refactoring tools from the original logic, and it produces better IL, which is why i used it.